### PR TITLE
DE-3707: pypi publish GitHub action POC

### PR DESCRIPTION
**Pull request by setiat**
## JIRA ticket
https://tetrascience.atlassian.net/browse/DE-3707

## Description
Making this PR a draft PR as it is a POC

- Check if release is triggered by authorised user
- Check if build version matches release version
- Build and publish using poetry if Poetry.lock is present
- Else use setup.py to build
- fail is both setup.py and poetry.lock are present

## Automated testing

- [ ] Unit tests
- [ ] Integration tests
- [ ] API tests
- [x] End-To-End tests
- [ ] N/A (please clarify)

Invalid Version: https://github.com/tetrascience/ts-pypi-integ-test/runs/6587439287?check_suite_focus=true

Publish with setup.py: https://github.com/tetrascience/ts-pypi-integ-test/runs/6587476238?check_suite_focus=true
Publish with poetry: https://github.com/tetrascience/ts-pypi-integ-test/runs/6587541431?check_suite_focus=true

Unauthorized Github actor: https://github.com/tetrascience/ts-pypi-integ-test/runs/6587932120?check_suite_focus=true

### CODE REVIEW

File: .github/workflows/pypi_publish.yml [hunk 1]
```diff
@@ -0,0 +1,127 @@
+# This workflow will add 2 jobs:
+# - check-format: checks whether any files would be reformatted by `black`
+# - test: runs `pytest`.  Uses `pipenv`, `poetry`, or `pip` depending on the files present
+
+name: PyPI Publisher
+
+on:
+  workflow_call:
+    inputs:
+      # Python version is required in order to make sure that pytest and black can be run correctly
```
Comment (abclay) [inline .github/workflows/pypi_publish.yml:10]:
```suggestion
      # Python version is required in order to make sure that the package can be built
```

File: .github/workflows/pypi_publish.yml [hunk 2]
```diff
@@ -0,0 +1,127 @@
+# This workflow will add 2 jobs:
+# - check-format: checks whether any files would be reformatted by `black`
+# - test: runs `pytest`.  Uses `pipenv`, `poetry`, or `pip` depending on the files present
```
[the remark this answers is not in the document]
Comment (setiat) [reply .github/workflows/pypi_publish.yml]: removed

File: .github/workflows/pypi_publish.yml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+          done < $repo_file
+
+          if [[ $authorized == false ]]; then
+              echo "Unauthorized user: $GITHUB_ACTOR"
+              echo "Authorized Users:"
+              echo `cat $repo_file`
+              exit 1
+          fi
+
+      - name: Validate Package Manger
```
[the remark this answers is not in the document]
Comment (setiat) [reply .github/workflows/pypi_publish.yml]: fixed

File: .github/workflows/pypi_publish.yml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+          while read line; do
+              if [[ $authorized == true ]]; then
+                  break
+              fi
+
+              if [[ $line == $GITHUB_ACTOR ]]; then
+                  echo "$line==$GITHUB_ACTOR"
+                  authorized=true
+              fi
+          done < $repo_file
```
[the remark this answers is not in the document]
Comment (setiat) [reply .github/workflows/pypi_publish.yml]: updated

File: .github/workflows/pypi_publish.yml [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+          echo "\nValidating Build Version"
+          build_version=$(ls dist/*tar.gz | awk -F"-" '{ print $2 }' | awk -F".tar" '{ print $1 }')
+          github_release_version=$(echo $GITHUB_REF_NAME | tr -d 'v')
+          echo "Build Version: ${build_version}"
+          echo "Github Release version: ${github_release_version}"
+          if [[ "$build_version" != "$github_release_version" ]]; then
+            echo "Release Version mismatch. Stopping..."
+            exit 1
+          fi
```
[the remark this answers is not in the document]
Comment (setiat) [reply .github/workflows/pypi_publish.yml]: updated

File: .github/workflows/pypi_publish.yml (hunk 1 above)
Comment (setiat) [reply .github/workflows/pypi_publish.yml]: updated

File: .github/workflows/pypi_publish.yml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    name: PyPI Publish
+    runs-on: ubuntu-latest
+    steps:
+      # Checkout the code
+      - uses: actions/checkout@v2
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ inputs.python-version }}
+      - name: Validate github actor
```
[the remark this answers is not in the document]
Comment (setiat) [reply .github/workflows/pypi_publish.yml:29]: added

**Comment by VladimirShteyn** [comment]
Can we close this ticket? It's already been implemented as part of DEI-185.

**Comment by abclay** [comment]
> Can we close this ticket? It's already been implemented as part of DEI-185.

I think so.  @setiat: any objections?

**Pull request closed without merging**